consensus/src/chained_bft/liveness/leader_reputation.rs

```diff
@@ -0,0 +1,102 @@
+// Copyright (c) The Libra Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::chained_bft::liveness::proposer_election::ProposerElection;
+use consensus_types::{
+    block::Block,
+    common::{Author, Round},
+};
+use libra_types::block_metadata::BlockMetadata;
+use rand::{
+    distributions::{Distribution, WeightedIndex},
+    rngs::StdRng,
+    SeedableRng,
+};
+use serde::export::PhantomData;
+use std::{collections::HashSet, sync::Arc};
+
+pub trait MetadataBackend {
+    fn get_meta_data(&self, window_size: u64) -> &[BlockMetadata];
```

Comment:
There is an implicit assumption about the end round of the window, which is seems wrong. When this function is called from `LeaderReputation::get_valid_proposers(round)`, how do we make sure the round of the metadata backend corresponds to the round of the function?
Do you think we should have `get_metadata(&self, window_size: u64, end_round: Round)`?.
Can we also add the comments regarding the returned vector? (what's the size of the vector in the beginning of the epoch?)

Reply:
Good point, it's intentional left as this, I should have left some comments on the behavior of this function.

It basically returns a slice ending with last committed block metadata with at most window_size items. So it'd be 0 in the beginning of the epoch, although we have to add epoch to metadata in order to filter it out.

It implicitly uses 0 as the gap we discussed, but I think it should be enough if we process sync_info of every message first.

It'd work as long as the proposer has highest commit and is honest, and proposer is likely to have highest commit because it aggregates the votes (which has 2f+1 sync_info).

Reply:
I still suspect it might not work.
Consider the happy path.
* Honest voters at round `10` have their last committed block at round `7`, they calculate the aggregator via `get_valid_proposers(11)` based on the metadata of rounds `[0-7]`.
* When the aggregator successfully aggregates the QC, it commits block at round `8`, hence when it moves to round `11` it calculates `get_valid_proposers(11)` based on the metadata of rounds `[0-8]`.

I might miss something obvious, but it seems the right way would be to take the `round` parameter into account and to consider only the committed metadata up to `round - 4` (inclusive)

Reply:
ah, I didn't think about this one-round off scenarios, thanks for pointing it out!